### ISSUE TITLE
Document behavior of #[cfg] in macros

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/reference#836]
The [macro reference](https://doc.rust-lang.org/reference/macros-by-example.html) should explain how `#[cfg]` and `#[cfg_attr(...)]` behavesinside macros.

Does this macro check the `allow_cfg` flag of the crate the macro is defined in, or of the crate the macro is *used* in?
```rust
macro_rules! field_check {
    ($type:path, $field:tt) => {
        // Make sure the field actually exists. This line ensures that a
        // compile-time error is generated if $field is accessed through a
        // Deref impl.
        #[cfg_attr(allow_clippy, allow(clippy::unneeded_field_pattern))]
        let $type { $field: _, .. };
    };
}
```

**Comment by ehuss** [comment]
I assume you are interested in more than just allow/deny diagnostic behavior. I mention that because I believe (some?) diagnostics have special behavior in external macros where [they are disabled](https://github.com/rust-lang/rust/blob/1a4e2b6f9c75a0e21722c88a0e3b610d6ffc3ae3/src/librustc_session/lint.rs#L82-L83).  The reference has historically shied away from documenting diagnostics, so that particular behavior is a bit of a gray area, but probably should be mentioned.

There are a few other behaviors of `cfg`/`cfg_attr` that I'd like to see documented (#565, #103).  In particular, I think the order of transformations ("phases of translation" in C++ parlance) would be useful to specify to make it clear *when* `cfg` processing is done in relation to other phases.

**Comment by RalfJung** [comment]
> I assume you are interested in more than just allow/deny diagnostic behavior.

Yes, I wonder about `cfg` in general (including `cfg!`). I didn't even know diagnostics have some special magic hacks.^^
